source/user/lmbench3/doc/lat_ctx.8
.\" $Id: lat_ctx.8,v 1.1.1.1 2006-11-23 11:25:34 steven Exp $
.TH LAT_CTX 8 "$Date: 2006-11-23 11:25:34 $" "(c)1994-2000 Carl Staelin and Larry McVoy" "LMBENCH"
.SH NAME
lat_ctx \- context switching benchmark
.SH SYNOPSIS
.B lat_ctx 
[
.I "-P <parallelism>"
]
[
.I "-W <warmups>"
]
[
.I "-N <repetitions>"
]
[
.I "-s <size_in_kbytes>"
]
.I "#procs"
[
.I "#procs ..."
]
.SH DESCRIPTION
.B lat_ctx
measures context switching time for any reasonable
number of processes of any reasonable size.
The processes are connected in a ring of Unix pipes.  Each process
reads a token from its pipe, possibly does some work, and then writes
the token to the next process.
.LP
Processes may vary in number.  Smaller numbers of processes result in
faster context switches.  More than 20 processes is not supported.
.LP
Processes may vary in size.  A size of zero is the baseline process that
does nothing except pass the token on to the next process.  A process size
of greater than zero means that the process does some work before passing
on the token.  The work is simulated as the summing up of an array of the
specified size.  The summing is an unrolled loop of about a 2.7 thousand
instructions.  
.LP
The effect is that both the data and the instruction cache
get polluted by some amount before the token is passed on.  The data 
cache gets polluted by approximately the process ``size''.  The instruction
cache gets polluted by a constant amount, approximately 2.7
thousand instructions.  
.LP
The pollution of the caches results in larger context switching times for
the larger processes.  This may be confusing because the benchmark takes
pains to measure only the context switch time, not including the overhead
of doing the work.  The subtle point is that the overhead is measured using
hot caches.  As the number and size of the processes increases, the caches
are more and more polluted until the set of processes do not fit.  The 
context switch times go up because a context switch is defined as the switch 
time
plus the time it takes to restore all of the process state, including 
cache state.  This means that the switch includes the time for the cache
misses on larger processes.
.SH OUTPUT
Output format is intended as input to \fBxgraph\fP or some similar program.
The format is multi line, the first line is a title that specifies the
size and non-context switching overhead of the test.  Each subsequent 
line is a pair of numbers that indicates the number of processes and 
the cost of a context switch.  The overhead and the context switch times are
in micro second units.  The numbers below are for a SPARCstation 2.
.sp
.ft CB
.nf
"size=0 ovr=179
2 71
4 104
8 134
16 333
20 438
.br
.fi
.ft
.SH BUGS
The numbers produced by this benchmark are somewhat inaccurate; they vary
by about 10 to 15% from run to run.  A series of runs may be done and the
lowest numbers reported.  The lower the number the more accurate the results.
.LP
The reasons for the inaccuracies are possibly interaction between the 
VM system and the processor caches.  It is possible that sometimes the
benchmark processes are laid out in memory such that there are fewer 
TLB/cache conflicts than other times.  This is pure speculation on our part.
.SH ACKNOWLEDGEMENT
Funding for the development of
this tool was provided by Sun Microsystems Computer Corporation.
.SH "SEE ALSO"
lmbench(8).
.SH "AUTHOR"
Carl Staelin and Larry McVoy
.PP
Comments, suggestions, and bug reports are always welcome.



source/History
Ralink AP SDK - History

Version 4.0.1.0
===============

- [IPv6] Avoid both eth2.2 and eth2 set same IPV6 address after receiving RA packet in gateway mode.
- [IPv6] Fix HNAT misbinding issue (RT6855/6)
- [IPv6] Fix ping fail(ping ::1) issue of loopback interface when we try to enable and disable disable_ipv6 parameter in /proc directory (RT6855/6)
- [Switch] Fix legacy switch(RT3052/RT3352/RT5350) settings vlan 12345 error
- [PPPoE] Kernel panic after PAP fail (RT6855/6)
- [PPTP] Kernel panic after PPTP peer terminated the link (RT6855/6)
- [SPI Flash] There are some noise if we try to program flash while I2S is playing music (RT3052/RT3883/RT3352/RT5350)
- [GPIO] Fix GPIO interrupt handler cannot be trigged correctly (RT6855/6)
- [GUI] Fix WPS bug for test plan
- [NET] Recover irqs_disabled() check under SKB Recycle to fix warning messages, but it decreases RT6855/6's maximum peek throughput
- [Uboot] RT6855/6 does watchdog reset at cold start to fix timer interrupt issue

Version 4.0.0.0
===============

PLATFORM:

- support RT3052/RT3883/RT3352/RT5350 (linux-2.6.21.x)
- support RT6855/RT6856 (linux-2.6.36MT.x)

NEW FUNCTIONS:

- [IPv6 ready logo Phase II] RT6855/RT6856
- [IPv6] support IPv6 Rapid Deployment
- [IPv6] support IPv6 DS-Lite
- [GUI] support two giga phy port
- [GUI] display AP Client site_survey

IMPROVEMENT:

- [IPv6] support ipv6 tftp server
- [PPTP/L2TP] performace tunning at Linux 2.6.21 (PPTP improve 25%, L2TP improve 10% compare to SDK 3.6)
- [SAMBA] significant performance fine tune on full loading scenario (dual-band concurrent AP with USB access)
- [NAND Flash] config page size and address cycle based on chip mode
- [I2S] Add MMAP for user/kernel
- [GUI] support WPS 2.0 

Bug fixed:

- [Kernel NVRAM] memory leakage / initiate failed / free cache / wirte data exceed 1024
- [RAETH] fix STP drop issue on RT3052 chipst ONLY
- [RAETH] fix switch abnormal after ifconfig down&up
- [RAETH] fix PDMA hang issue under heavy load test
- [RAETH] fix TSO bug if USB storage is enabled
- [USB device] system will hang whiling loading gadget driver. Turning off power-saving on both port0 and port1
- [USB device] force host to do re-enumeration to solve "unrecognize device" when re-plug-in in windows platform.
- [IPv6] kernel panic bug if enable ipv6 privacy extension or router prereference
- [IPv6] fix NS packet loss for ipv6 DAD
- [GUI] flush conntrack table while restart/deleting rules
- [NAND Flash] fix nand flash not work for MTD_ANY mode
- [IGMPPROXY] fixed the overflow FATAL bug while handling bit47=1 mac address
- [GPIO] fixed RT3352/RT5350 extended GPIO pins mapping



source/user/lmbench3/src/lat_ctx.c
/*
 * lat_ctx.c - context switch timer 
 *
 * usage: lat_ctx [-P parallelism] [-W <warmup>] [-N <repetitions>] [-s size] #procs [#procs....]
 *
 * Copyright (c) 1994 Larry McVoy.  Distributed under the FSF GPL with
 * additional restriction that results may published only if
 * (1) the benchmark is unmodified, and
 * (2) the version in the sccsid below is included in the report.
 * Support for this development by Sun Microsystems is gratefully acknowledged.
 */
char	*id = "$Id: lat_ctx.c,v 1.1.1.1 2006-11-23 11:25:34 steven Exp $\n";

#include "bench.h"


#define	MAXPROC	2048
#define	CHUNK	(4<<10)
#define	TRIPS	5
#ifndef	max
#define	max(a, b)	((a) > (b) ? (a) : (b))
#endif

void	doit(int rd, int wr, int process_size);
int	create_pipes(int **p, int procs);
int	create_daemons(int **p, pid_t *pids, int procs, int process_size);
void	initialize_overhead(iter_t iterations, void* cookie);
void	cleanup_overhead(iter_t iterations, void* cookie);
void	benchmark_overhead(iter_t iterations, void* cookie);
void	initialize(iter_t iterations, void* cookie);
void	cleanup(iter_t iterations, void* cookie);
void	benchmark(iter_t iterations, void* cookie);

struct _state {
	int	process_size;
	double	overhead;
	int	procs;
	pid_t*	pids;
	int	**p;
	void*	data;
};

int
main(int ac, char **av)
{
	int	i, maxprocs;
	int	c;
	int	parallel = 1;
	int	warmup = 0;
	int	repetitions = TRIES;
	struct _state state;
	char *usage = "[-P <parallelism>] [-W <warmup>] [-N <repetitions>] [-s kbytes] processes [processes ...]\n";
	double	time;

	/*
	 * Need 4 byte ints.
	 */
	if (sizeof(int) != 4) {
		fprintf(stderr, "Fix sumit() in ctx.c.\n");
		exit(1);
	}

	state.process_size = 0;
	state.overhead = 0.0;
	state.pids = NULL;

	/*
	 * If they specified a context size, or parallelism level, get them.
	 */
	while (( c = getopt(ac, av, "s:P:W:N:")) != EOF) {
		switch(c) {
		case 'P':
			parallel = atoi(optarg);
			if (parallel <= 0) lmbench_usage(ac, av, usage);
			break;
		case 'W':
			warmup = atoi(optarg);
			break;
		case 'N':
			repetitions = atoi(optarg);
			break;
		case 's':
			state.process_size = atoi(optarg) * 1024;
			break;
		default:
			lmbench_usage(ac, av, usage);
			break;
		}
	}

	if (optind > ac - 1)
		lmbench_usage(ac, av, usage);

	/* compute pipe + sumit overhead */
	maxprocs = atoi(av[optind]);
	for (i = optind; i < ac; ++i) {
		state.procs = atoi(av[i]);
		if (state.procs > maxprocs)
			maxprocs = state.procs;
	}
	state.procs = maxprocs;
	benchmp(initialize_overhead, benchmark_overhead, cleanup_overhead, 
		0, 1, warmup, repetitions, &state);
	if (gettime() == 0) return(0);
	state.overhead = gettime();
	state.overhead /= get_n();
	fprintf(stderr, "\n\"size=%dk ovr=%.2f\n", 
		state.process_size/1024, state.overhead);

	/* compute the context switch cost for N processes */
	for (i = optind; i < ac; ++i) {
		state.procs = atoi(av[i]);
		benchmp(initialize, benchmark, cleanup, 0, parallel, 
			warmup, repetitions, &state);

		time = gettime();
		time /= get_n();
		time /= state.procs;
		time -= state.overhead;

		if (time > 0.0)
			fprintf(stderr, "%d %.2f\n", state.procs, time);
	}

	return (0);
}

void
initialize_overhead(iter_t iterations, void* cookie)
{
	int i;
	int procs;
	int* p;
	struct _state* pState = (struct _state*)cookie;

	if (iterations) return;

	pState->pids = NULL;
	pState->p = (int**)malloc(pState->procs * (sizeof(int*) + 2 * sizeof(int)));
	p = (int*)&pState->p[pState->procs];
	for (i = 0; i < pState->procs; ++i) {
		pState->p[i] = p;
		p += 2;
	}

	pState->data = (pState->process_size > 0) ? malloc(pState->process_size) : NULL;
	if (pState->data)
		bzero(pState->data, pState->process_size);

	procs = create_pipes(pState->p, pState->procs);
	if (procs < pState->procs) {
		cleanup_overhead(0, cookie);
		exit(1);
	}
}

void
cleanup_overhead(iter_t iterations, void* cookie)
{
	int i;
	struct _state* pState = (struct _state*)cookie;

	if (iterations) return;

     	for (i = 0; i < pState->procs; ++i) {
		close(pState->p[i][0]);
		close(pState->p[i][1]);
	}

	free(pState->p);
	if (pState->data) free(pState->data);
}

void
benchmark_overhead(iter_t iterations, void* cookie)
{
	struct _state* pState = (struct _state*)cookie;
	int	i = 0;
	int	msg = 1;

	while (iterations-- > 0) {
		if (write(pState->p[i][1], &msg, sizeof(msg)) != sizeof(msg)) {
			/* perror("read/write on pipe"); */
			exit(1);				
		}
		if (read(pState->p[i][0], &msg, sizeof(msg)) != sizeof(msg)) {
			/* perror("read/write on pipe"); */
			exit(1);
		}
		if (++i == pState->procs) {
			i = 0;
		}
		bread(pState->data, pState->process_size);
	}
}

void 
initialize(iter_t iterations, void* cookie)
{
	int procs;
	struct _state* pState = (struct _state*)cookie;

	if (iterations) return;

	initialize_overhead(iterations, cookie);

	pState->pids = (pid_t*)malloc(pState->procs * sizeof(pid_t));
	if (pState->pids == NULL)
		exit(1);
	bzero((void*)pState->pids, pState->procs * sizeof(pid_t));
	procs = create_daemons(pState->p, pState->pids, 
			       pState->procs, pState->process_size);
	if (procs < pState->procs) {
		cleanup(0, cookie);
		exit(1);
	}
}

void
cleanup(iter_t iterations, void* cookie)
{
	int i;
	struct _state* pState = (struct _state*)cookie;

	if (iterations) return;

	/*
	 * Close the pipes and kill the children.
	 */
	cleanup_overhead(iterations, cookie);
     	for (i = 1; pState->pids && i < pState->procs; ++i) {
		if (pState->pids[i] > 0) {
			kill(pState->pids[i], SIGKILL);
			waitpid(pState->pids[i], NULL, 0);
		}
	}
	if (pState->pids)
		free(pState->pids);
	pState->pids = NULL;
}

void
benchmark(iter_t iterations, void* cookie)
{
	struct _state* pState = (struct _state*)cookie;
	int	msg;

	/*
	 * Main process - all others should be ready to roll, time the
	 * loop.
	 */
	while (iterations-- > 0) {
		if (write(pState->p[0][1], &msg, sizeof(msg)) !=
		    sizeof(msg)) {
			/* perror("read/write on pipe"); */
			exit(1);
		}
		if (read(pState->p[pState->procs-1][0], &msg, sizeof(msg)) != sizeof(msg)) {
			/* perror("read/write on pipe"); */
			exit(1);
		}
		bread(pState->data, pState->process_size);
	}
}


void
doit(int rd, int wr, int process_size)
{
	int	msg;
	void*	data = NULL;

	if (process_size) {
		data = malloc(process_size);
		if (data) bzero(data, process_size);
	}
	for ( ;; ) {
		if (read(rd, &msg, sizeof(msg)) != sizeof(msg)) {
			/* perror("read/write on pipe"); */
			break;
		}
		bread(data, process_size);
		if (write(wr, &msg, sizeof(msg)) != sizeof(msg)) {
			/* perror("read/write on pipe"); */
			break;
		}
	}
	exit(1);
}


int
create_daemons(int **p, pid_t *pids, int procs, int process_size)
{
	int	i, j;
	int	msg;

	/*
	 * Use the pipes as a ring, and fork off a bunch of processes
	 * to pass the byte through their part of the ring.
	 *
	 * Do the sum in each process and get that time before moving on.
	 */
	handle_scheduler(benchmp_childid(), 0, procs-1);
     	for (i = 1; i < procs; ++i) {
		switch (pids[i] = fork()) {
		    case -1:	/* could not fork, out of processes? */
			return i;

		    case 0:	/* child */
			handle_scheduler(benchmp_childid(), i, procs-1);
			for (j = 0; j < procs; ++j) {
				if (j != i - 1) close(p[j][0]);
				if (j != i) close(p[j][1]);
			}
			doit(p[i-1][0], p[i][1], process_size);
			/* NOTREACHED */

		    default:	/* parent */
			;
	    	}
	}

	/*
	 * Go once around the loop to make sure that everyone is ready and
	 * to get the token in the pipeline.
	 */
	if (write(p[0][1], &msg, sizeof(msg)) != sizeof(msg) ||
	    read(p[procs-1][0], &msg, sizeof(msg)) != sizeof(msg)) {
		/* perror("write/read/write on pipe"); */
		exit(1);
	}
	return procs;
}

int
create_pipes(int **p, int procs)
{
	int	i;
	/*
	 * Get a bunch of pipes.
	 */
	morefds();
     	for (i = 0; i < procs; ++i) {
		if (pipe(p[i]) == -1) {
			return i;
		}
	}
	return procs;
}
